File: src/domain/models/hand.ts
import { Card, Rank, rankBjValue } from './card'

export class Hand {
  private _cards: Card[] = []

  get cards(): readonly Card[] { return this._cards }

  add(card: Card): void { this._cards.push(card) }
  clear(): void { this._cards = [] }

  // Generates every possible score combination for all Ace variants via reduce+flatMap
  private get allScores(): number[] {
    return this._cards.reduce<number[]>((totals, card) => {
      const values = rankBjValue(card.rank)
      return values.flatMap(v => totals.map(t => t + v))
    }, [0])
  }

  get bestScore(): number {
    const valid = this.allScores.filter(s => s <= 21)
    if (valid.length > 0) return Math.max(...valid)
    const all = this.allScores
    return all.length > 0 ? Math.min(...all) : 0
  }

  // Hard total: every Ace counts as 1
  get hardTotal(): number {
    return this._cards.reduce((sum, card) => {
      return sum + (card.rank === Rank.ACE ? 1 : rankBjValue(card.rank)[0])
    }, 0)
  }

  get isSoft(): boolean {
    if (!this._cards.some(c => c.rank === Rank.ACE)) return false
    return this.bestScore <= 21 && this.bestScore !== this.hardTotal
  }

  get isBust(): boolean { return this.bestScore > 21 }

  get isBlackjack(): boolean {
    return this._cards.length === 2 && this.bestScore === 21
  }

  get isPair(): boolean {
    return this._cards.length === 2 && this._cards[0]!.rank === this._cards[1]!.rank
  }

  get isTerminal(): boolean { return this.isBust || this.bestScore === 21 }
}

File: BlackjackApp/Domain/Models/Hand.swift
import Foundation

struct Hand: Sendable {
    private(set) var cards: [Card] = []

    // MARK: – Mutation

    mutating func add(_ card: Card) {
        cards.append(card)
    }

    mutating func clear() {
        cards.removeAll()
    }

    // MARK: – Score computation

    /// All possible totals (before pruning busts)
    private var allScores: [Int] {
        cards.reduce([0]) { totals, card in
            card.rank.bjValue.flatMap { value in
                totals.map { $0 + value }
            }
        }
    }

    /// Best non-bust score; if all options bust, returns lowest possible total.
    var bestScore: Int {
        let valid = allScores.filter { $0 <= 21 }
        return valid.max() ?? allScores.min() ?? 0
    }

    /// Hard total: treating every Ace as 1 only.
    var hardTotal: Int {
        cards.reduce(0) { $0 + ($1.rank == .ace ? 1 : $1.rank.bjValue[0]) }
    }

    /// True when the best non-bust score uses an Ace counted as 11.
    var isSoft: Bool {
        guard cards.contains(where: { $0.rank == .ace }) else { return false }
        let score = bestScore
        return score <= 21 && score != hardTotal
    }

    var isBust: Bool { bestScore > 21 }

    var isBlackjack: Bool { cards.count == 2 && bestScore == 21 }

    /// True when the initial two cards share the same rank (split candidate).
    var isPair: Bool { cards.count == 2 && cards[0].rank == cards[1].rank }

    /// Convenience: is this hand done (bust, 21, or blackjack)?
    var isTerminal: Bool { isBust || bestScore == 21 }
}
